oggdec/oggdec.1
.TH "oggdec" "1" "2008 September 9" "Xiph.Org Foundation" "Vorbis Tools"

.SH "NAME"
oggdec - simple decoder, Ogg Vorbis file to PCM audio file (Wave or RAW).

.SH "SYNOPSIS"
.B oggdec
[
.B -Qhv
] [
.B -b bits_per_sample
] [
.B -e endianness
] [
.B -R
] [
.B -s signedness
] [
.B -o outputfile
] 
.B file ...

.SH "DESCRIPTION"

.B oggdec
decodes Ogg Vorbis files into PCM-encoded ("uncompressed") audio files, either Wave or RAW format.

For each input file, 
.B oggdec
writes to a filename based on the input filename, but with the extension changed to ".wav" or ".raw" as appropriate.

If the input file is specified as
.B "-"
, then 
.B oggdec
will read from stdin, and write to stdout unless an output filename is specified. Likewise, an output filename of
.B -
will cause output to be to stdout.

Writing Wave format to stdout is a bad idea.  Wave requires a seekable medium for the header to be rewritten after all the data is written out; stdout is not seekable.

.SH "OPTIONS"
.IP "-Q, --quiet"
Suppresses program output.
.IP "-h, --help"
Print help message.
.IP "-V, --version"
Display version information.
.IP "-b n, --bits=n"
Bits per sample.  Valid values are 8 or 16.
.IP "-e n, --endian=n"
Set endianness for 16-bit output.  0 (default) is little-endian (Intel byte order).  1 is big-endian (sane byte order).
.IP "-R, --raw"
Output in raw format.  If not specified, writes Wave file (RIFF headers).
.IP "-s n, --sign=n"
Set signedness for output.  0 for unsigned, 1 (default) for signed.
.IP "-o filename, --output=filename"
Write output to specified filename.  This option is only valid if one input [file] is specified, or if raw mode is used.

.SH "EXAMPLES"
Decode a file 
.B enabler.ogg
to 
.B enabler.wav
 as little-endian signed 16-bit (default options):
.RS
oggdec enabler.ogg
.RE

Decode a file 
.B enabler.ogg
to 
.B enabler.raw
as headerless little-endian signed 16-bit:
.RS
oggdec --raw enabler.ogg
.RE

Decode 
.B enabler.ogg
to 
.B enabler.crazymonkey
as unsigned 8-bit:
.RS
oggdec -b 8 -s 0 -o enabler.crazymonkey enabler.ogg
.RE

Decode 
.B enabler.ogg
to 
.B enabler.raw
as big-endian signed 16-bit (any of the following):
.RS
oggdec -R -e 1 -b 16 enabler.ogg
.RE
.RS
oggdec -R -e 1 -b 16 -o enabler.raw - < enabler.ogg
.RE
.RS
oggdec -R -e 1 -b 16 - < enabler.ogg > enabler.raw
.RE

Mass decoding (foo.ogg to foo.wav, bar.ogg to bar.wav, quux.ogg to quux.wav, etc.):
.RS
oggdec *.ogg
.RE

.SH "AUTHORS"
.SS "Program Authors"
Michael Smith <msmith@xiph.org>
.SS "Manpage Authors"

.br

Frederick Lee <phaethon@linux.ucla.edu>, assisted by a few million monkeys armed with keyboards in irc://irc.openprojects.net/#vorbis

.SH "SEE ALSO"

.PP
\fBogg123\fR(1), \fBoggenc\fR(1), \fBvorbiscomment\fR(1), \fBflac\fR(1), \fBspeexdec\fR(1)


CHANGES
vorbis-tools 1.3.0 -- Unreleased

 * Fixed an error in configure.ac; --with-speex/flac work again (#1319)
 * Corrected problems in the Visual Studio project files
 * Updated po files from the Translation Project
 * Added new en_GB.po, eo.po, pl.po, sk.po and vi.po translation files
 * Added AC_USE_SYSTEM_EXTENSIONS to configure.ac; no more configure warnings
 * Corrected SUBLANG values in intl/localename.c (#1415)
 * Change -v to -V on oggenc and oggdec for consistency (#1112)
 * Fix for utf8_decode in Windows; improves behavior in vorbiscomment (#268)
 * Updated gettextize to version 0.17
 * ogg123: backported fix from libfishsound to patch the Speex decoder (#1347)
 * ogg123: fixed CPU issue when outputting to a closed pipe (#1357)
 * ogg123: return value to stop decoding after buffer is shut down (#1357)
 * ogg123: support for ReplayGain; requires libvorbis 1.2.1 or later (#381)
 * ogg123: unset non-blocking mode on stderr
 * oggdec: gettextized help text (#1385)
 * oggdec: gettextized all strings
 * oggdec: call ov_open_callbacks instead of ov_open; it works on Windows now
 * oggenc: fixed a core dump while resampling from FLAC (#1316)
 * oggenc: fixed a typo in the Skeleton handling routine
 * oggenc: fixed remapping channels bug (#1326)
 * oggenc: support for WAVE_FORMAT_EXTENSIBLE headers (#1326)
 * oggenc: support for 32 bit Wave files (#1326)
 * oggenc: --ignorelength; support for Wave files > 4 GB (#1326)
 * oggenc: split help text into manageable chunks to help translators (#1385)
 * oggenc: --utf8 command similar to vorbiscomment's --raw (#268)
 * oggenc: fixed the encoding of extended characters in Windows (#268)
 * oggenc: validate raw UTF-8 sent to oggenc (#268)
 * oggenc: include the PID in the RNG seed to get a unique serial (#1432)
 * oggenc: lyrics support using .lrc as source; requires libkate (#1403)
 * ogginfo: support for information in Kate streams (#1360)
 * vcut: 64 bit fixes (#1366)
 * vcut: support for chained streams (#1455)
 * vorbiscomment: correct memory allocation (#472)
 * vorbiscomment: validate raw UTF-8 sent to vorbiscomment (#268)
 * vorbiscomment: fix segfault when using --tag (#1439)
 * vorbiscomment: round-trip multi-line comments (#273)

vorbis-tools 1.2.0 -- 2008-02-21

 * FLAC support now relies solely on libFLAC
 * Support for large files (#879)
 * Fixed acinclude.m4 to properly support --no switches
 * ogg123: added remote control support (#1109)
 * ogg123: fixed a bug in esd when pressing CTRL + C (#715)
 * ogg123: fixed a type mismatch in option parsing for 64 bit systems
 * ogg123: configuration no longer hardcoded to /etc
 * ogg123: compiles with older versions of libcurl
 * ogg123: fixed crash when playing 1-channel FLAC (#535)
 * ogg123: fixed floating-point exception when playing corrupt FLAC (#1119)
 * oggdec: limited support for chained Ogg bitstreams
 * oggdec: support decoding of multiple files into a single one
 * oggenc: -k, switch for Skeleton bitstream encoding
 * oggenc: fixed issues with Skeleton on big endian systems
 * oggenc: proper 5.1 channel mapping support
 * oggenc: FLAC module does not confuse every Ogg file as its own
 * oggenc: compiles with older versions of libvorbis
 * ogginfo: recognizes Skeleton, Dirac, FLAC and Kate bitstreams
 * vcut: solved issues described in ticket #1313
 * vorbiscomment: support for creation of long comments
 * vorbiscomment: support for multiplexed Vorbis
 * Several minor bug fixes

oggdec/oggdec.c
/* OggDec
 *
 * This program is distributed under the GNU General Public License, version 2.
 * A copy of this license is included with this source.
 *
 * Copyright 2002, Michael Smith <msmith@xiph.org>
 *
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdlib.h>
#include <stdio.h>
#include <getopt.h>
#include <errno.h>
#include <string.h>
#include <locale.h>

#if defined(_WIN32) || defined(__EMX__) || defined(__WATCOMC__)
#include <fcntl.h>
#include <io.h>
#endif

#include <vorbis/vorbisfile.h>

#include "i18n.h"

struct option long_options[] = {
    {"quiet", no_argument, NULL, 'Q'},
    {"help", no_argument, NULL, 'h'},
    {"version", no_argument, NULL, 'V'},
    {"bits", required_argument, NULL, 'b'},
    {"endianness", required_argument, NULL, 'e'},
    {"raw", no_argument, NULL, 'R'},
    {"sign", required_argument, NULL, 's'},
    {"output", required_argument, NULL, 'o'},
    {NULL,0,NULL,0}
};

static int quiet = 0;
static int bits = 16;
static int endian = 0;
static int raw = 0;
static int sign = 1;
unsigned char headbuf[44]; /* The whole buffer */
char *outfilename = NULL;

static void version (void) {
    fprintf(stderr, _("oggdec from %s %s\n"), PACKAGE, VERSION);
}

static void usage(void)
{
    version ();
    fprintf(stdout, _(" by the Xiph.Org Foundation (https://www.xiph.org/)\n"));
    fprintf(stdout, _(" using decoder %s.\n\n"), vorbis_version_string());
    fprintf(stdout, _("Usage: oggdec [options] file1.ogg [file2.ogg ... fileN.ogg]\n\n"));
    fprintf(stdout, _("Supported options:\n"));
    fprintf(stdout, _(" --quiet, -Q      Quiet mode. No console output.\n"));
    fprintf(stdout, _(" --help,  -h      Produce this help message.\n"));
    fprintf(stdout, _(" --version, -V    Print out version number.\n"));
    fprintf(stdout, _(" --bits, -b       Bit depth for output (8 and 16 supported)\n"));
    fprintf(stdout, _(" --endianness, -e Output endianness for 16-bit output; 0 for\n"
                      "                  little endian (default), 1 for big endian.\n"));
    fprintf(stdout, _(" --sign, -s       Sign for output PCM; 0 for unsigned, 1 for\n"
                      "                  signed (default 1).\n"));
    fprintf(stdout, _(" --raw, -R        Raw (headerless) output.\n"));
    fprintf(stdout, _(" --output, -o     Output to given filename. May only be used\n"
                      "                  if there is only one input file, except in\n"
                      "                  raw mode.\n"));
}

static void parse_options(int argc, char **argv)
{
    int option_index = 1;
    int ret;

    while((ret = getopt_long(argc, argv, "QhVb:e:Rs:o:", 
                    long_options, &option_index)) != -1)
    {
        switch(ret)
        {
            case 'Q':
                quiet = 1;
                break;
            case 'h':
                usage();
                exit(0);
                break;
            case 'V':
                version();
                exit(0);
                break;
            case 's':
                sign = atoi(optarg);
                break;
            case 'b':
                bits = atoi(optarg);
                if(bits <= 8)
                    bits = 8;
                else
                    bits = 16;
                break;
            case 'e':
                endian = atoi(optarg);
                break;
            case 'o':
                outfilename = strdup(optarg);
                break;
            case 'R':
                raw = 1;
                break;
            default:
                fprintf(stderr, _("Internal error: Unrecognised argument\n"));
                break;
        }
    }
}

#define WRITE_U32(buf, x) *(buf)     = (unsigned char)((x)&0xff);\
                          *((buf)+1) = (unsigned char)(((x)>>8)&0xff);\
                          *((buf)+2) = (unsigned char)(((x)>>16)&0xff);\
                          *((buf)+3) = (unsigned char)(((x)>>24)&0xff);

#define WRITE_U16(buf, x) *(buf)     = (unsigned char)((x)&0xff);\
                          *((buf)+1) = (unsigned char)(((x)>>8)&0xff);

/* Some of this based on ao/src/ao_wav.c */
int write_prelim_header(OggVorbis_File *vf, FILE *out, ogg_int64_t knownlength) {
    unsigned int size = 0x7fffffff;
    int channels = ov_info(vf,0)->channels;
    int samplerate = ov_info(vf,0)->rate;
    int bytespersec = channels*samplerate*bits/8;
    int align = channels*bits/8;
    int samplesize = bits;

    if(knownlength && knownlength*bits/8*channels < size)
        size = (unsigned int)(knownlength*bits/8*channels+44) ;

    memcpy(headbuf, "RIFF", 4);
    WRITE_U32(headbuf+4, size-8);
    memcpy(headbuf+8, "WAVE", 4);
    memcpy(headbuf+12, "fmt ", 4);
    WRITE_U32(headbuf+16, 16);
    WRITE_U16(headbuf+20, 1); /* format */
    WRITE_U16(headbuf+22, channels);
    WRITE_U32(headbuf+24, samplerate);
    WRITE_U32(headbuf+28, bytespersec);
    WRITE_U16(headbuf+32, align);
    WRITE_U16(headbuf+34, samplesize);
    memcpy(headbuf+36, "data", 4);
    WRITE_U32(headbuf+40, size - 44);

    if(fwrite(headbuf, 1, 44, out) != 44) {
        fprintf(stderr, _("ERROR: Failed to write Wave header: %s\n"), strerror(errno));
        return 1;
    }

    return 0;
}

int rewrite_header(FILE *out, unsigned int written) 
{
    unsigned int length = written;

    length += 44;

    WRITE_U32(headbuf+4, length-8);
    WRITE_U32(headbuf+40, length-44);
    if(fseek(out, 0, SEEK_SET) != 0)
        return 1;

    if(fwrite(headbuf, 1, 44, out) != 44) {
        fprintf(stderr, _("ERROR: Failed to write Wave header: %s\n"), strerror(errno));
        return 1;
    }
    return 0;
}

static FILE *open_input(char *infile) 
{
    FILE *in;

    if(!infile) {
#ifdef __BORLANDC__
        setmode(fileno(stdin), O_BINARY);
#elif _WIN32
        _setmode(_fileno(stdin), _O_BINARY);
#endif
        in = stdin;
    }
    else {
        in = fopen(infile, "rb");
        if(!in) {
            fprintf(stderr, _("ERROR: Failed to open input file: %s\n"), strerror(errno));
            return NULL;
        }
    }

    return in;
}

static FILE *open_output(char *outfile) 
{
    FILE *out;
    if(!outfile) {
#ifdef __BORLANDC__
        setmode(fileno(stdout), O_BINARY);
#elif _WIN32
        _setmode(_fileno(stdout), _O_BINARY);
#endif
        out = stdout;
    }
    else {
        out = fopen(outfile, "wb");
        if(!out) {
            fprintf(stderr, _("ERROR: Failed to open output file: %s\n"), strerror(errno));
            return NULL;
        }
    }

    return out;
}

static void
permute_channels(char *in, char *out, int len, int channels, int bytespersample)
{
    int permute[6][6] = {{0}, {0,1}, {0,2,1}, {0,1,2,3}, {0,1,2,3,4}, 
        {0,2,1,5,3,4}};
    int i,j,k;
    int samples = len/channels/bytespersample;

    /* Can't handle, don't try */
    if (channels > 6)
        return;

    for (i=0; i < samples; i++) {
        for (j=0; j < channels; j++) {
            for (k=0; k < bytespersample; k++) {
                out[i*bytespersample*channels + 
                    bytespersample*permute[channels-1][j] + k] = 
                    in[i*bytespersample*channels + bytespersample*j + k];
            }
        }
    }
}

static int decode_file(FILE *in, FILE *out, char *infile, char *outfile)
{
    OggVorbis_File vf;
    int bs = 0;
    char buf[8192], outbuf[8192];
    char *p_outbuf;
    int buflen = 8192;
    unsigned int written = 0;
    int ret;
    ogg_int64_t length = 0;
    ogg_int64_t done = 0;
    int size = 0;
    int seekable = 0;
    int percent = 0;
    int channels;
    int samplerate;

    if (ov_open_callbacks(in, &vf, NULL, 0, OV_CALLBACKS_DEFAULT) < 0) {
        fprintf(stderr, _("ERROR: Failed to open input as Vorbis\n"));
        fclose(in);
        return 1;
    }

    channels = ov_info(&vf,0)->channels;
    samplerate = ov_info(&vf,0)->rate;

    if(ov_seekable(&vf)) {
        int link;
        int chainsallowed = 0;
        for(link = 0; link < ov_streams(&vf); link++) {
            if(ov_info(&vf, link)->channels == channels && 
                    ov_info(&vf, link)->rate == samplerate)
            {
                chainsallowed = 1;
            }
        }

        seekable = 1;
        if(chainsallowed)
            length = ov_pcm_total(&vf, -1);
        else
            length = ov_pcm_total(&vf, 0);
        size = bits/8 * channels;
        if(!quiet)
            fprintf(stderr, _("Decoding \"%s\" to \"%s\"\n"), 
                    infile?infile:_("standard input"), 
                    outfile?outfile:_("standard output"));
    }

    if(!raw) {
        if(write_prelim_header(&vf, out, length)) {
            ov_clear(&vf);
            return 1;
        }
    }

    while((ret = ov_read(&vf, buf, buflen, endian, bits/8, sign, &bs)) != 0) {
        if(bs != 0) {
            vorbis_info *vi = ov_info(&vf, -1);
            if(channels != vi->channels || samplerate != vi->rate) {
                fprintf(stderr, _("Logical bitstreams with changing parameters are not supported\n"));
                break;
            }
        }

        if(ret == OV_HOLE) {
            if(!quiet) {
                fprintf(stderr, _("WARNING: hole in data (%d)\n"), ret);
            }
            continue;
        }
        else if(ret < 0) {
            if(!quiet) {
                fprintf(stderr, _("=== Vorbis library reported a stream error.\n"));
            }
            ov_clear(&vf);
            return 1;
        }

        if(channels > 2 && !raw) {
          /* Then permute! */
          permute_channels(buf, outbuf, ret, channels, bits/8);
          p_outbuf = outbuf;
        }
        else {
          p_outbuf = buf;
        }

        if(fwrite(p_outbuf, 1, ret, out) != ret) {
            fprintf(stderr, _("Error writing to file: %s\n"), strerror(errno));
            ov_clear(&vf);
            return 1;
        }

        written += ret;
        if(!quiet && seekable) {
            done += ret/size;
            if((double)done/(double)length * 200. > (double)percent) {
                percent = (int)((double)done/(double)length *200);
                fprintf(stderr, "\r\t[%5.1f%%]", (double)percent/2.);
            }
        }
    }

    if(seekable && !quiet)
        fprintf(stderr, "\n");

    if(!raw)
        rewrite_header(out, written); /* We don't care if it fails, too late */

    ov_clear(&vf);

    return 0;
}

int main(int argc, char **argv)
{
    int i;

    setlocale(LC_ALL, "");
    bindtextdomain(PACKAGE, LOCALEDIR);
    textdomain(PACKAGE);

    if(argc == 1) {
        usage();
        return 1;
    }

    parse_options(argc,argv);

    if(!quiet)
        version();

    if(optind >= argc) {
        fprintf(stderr, _("ERROR: No input files specified. Use -h for help\n"));
        return 1;
    }

    if(argc - optind > 1 && outfilename && !raw) {
        fprintf(stderr, _("ERROR: Can only specify one input file if output filename is specified\n"));
        return 1;
    }

    if(outfilename && raw) {
        FILE *infile, *outfile;
        char *infilename;

        if(!strcmp(outfilename, "-")) {
            free(outfilename);
            outfilename = NULL;
        }
        outfile = open_output(outfilename);

        if(!outfile)
            return 1;

        for(i=optind; i < argc; i++) {
            if(!strcmp(argv[i], "-")) {
                infilename = NULL;
                infile = open_input(NULL);
            }
            else {
                infilename = argv[i];
                infile = open_input(argv[i]);
            }

            if(!infile) {
                fclose(outfile);
                free(outfilename);
                return 1;
            }
            if(decode_file(infile, outfile, infilename, outfilename)) {
                fclose(outfile);
                return 1;
            }

        }

        fclose(outfile);
    }
    else {
        for(i=optind; i < argc; i++) {
            char *in, *out;
            FILE *infile, *outfile;

            if(!strcmp(argv[i], "-"))
                in = NULL;
            else
                in = argv[i];

            if(outfilename) {
                if(!strcmp(outfilename, "-"))
                    out = NULL;
                else
                    out = outfilename;
            }
            else {
                if(!strcmp(argv[i], "-")) {
                    out = NULL;
                }
                else {
                    char *end = strrchr(argv[i], '.');
                    end = end?end:(argv[i] + strlen(argv[i]) + 1);

                    out = malloc(strlen(argv[i]) + 10);
                    strncpy(out, argv[i], end-argv[i]);
                    out[end-argv[i]] = 0;
                    if(raw)
                        strcat(out, ".raw");
                    else
                        strcat(out, ".wav");
                }
            }

            infile = open_input(in);
            if(!infile) {
                if(outfilename)
                    free(outfilename);
                return 1;
            }
            outfile = open_output(out);
            if(!outfile) {
                fclose(infile);
                return 1;
            }

            if(decode_file(infile, outfile, in, out)) {
                fclose(outfile);
                return 1;
            }

            if(!outfilename && out)
                free(out);

            fclose(outfile);
        }
    }

    if(outfilename)
        free(outfilename);

    return 0;
}
